ownCloud/Client/Actions/ClientDirectoryPickerViewController.swift

```diff
@@ -0,0 +1,97 @@
+//
+//  ClientDirectoryPickerViewController.swift
+//  ownCloud
+//
+//  Created by Pablo Carrascal on 22/08/2018.
+//  Copyright © 2018 ownCloud GmbH. All rights reserved.
+//
+
+import UIKit
+import ownCloudSDK
+
+class ClientDirectoryPickerViewController: ClientQueryViewController {
+
+	private var selectButton: ThemeButton
+	private var completion: (OCItem) -> Void
+	private var cancelBarButton: UIBarButtonItem!
+
+	init(core inCore: OCCore, query inQuery: OCQuery, completion: @escaping (OCItem) -> Void) {
+		selectButton = ThemeButton()
+		self.completion = completion
+		super.init(core: inCore, query: inQuery)
+
+		Theme.shared.register(client: self)
+	}
+
+	required init?(coder aDecoder: NSCoder) {
+		fatalError("init(coder:) has not been implemented")
+	}
+
+	deinit {
+		Theme.shared.unregister(client: self)
+	}
+
+	override func viewDidLayoutSubviews() {
+		super.viewDidLayoutSubviews()
+		tableView.bringSubview(toFront: selectButton)
+	}
+
+	override func viewDidLoad() {
+		super.viewDidLoad()
+
+		// Select Button setup
+		selectButton.translatesAutoresizingMaskIntoConstraints = false
+		tableView.addSubview(selectButton)
+
+		NSLayoutConstraint.activate([
+			selectButton.leftAnchor.constraint(equalTo: tableView.safeAreaLayoutGuide.leftAnchor),
+			selectButton.rightAnchor.constraint(equalTo: tableView.safeAreaLayoutGuide.rightAnchor),
+			selectButton.bottomAnchor.constraint(equalTo: tableView.safeAreaLayoutGuide.bottomAnchor),
+			selectButton.heightAnchor.constraint(equalToConstant: 44)
+		])
+
+		cancelBarButton = UIBarButtonItem(barButtonSystemItem: .cancel, target: self, action: #selector(cancelBarButtonPressed))
+		navigationItem.rightBarButtonItem = cancelBarButton
```

Comment:
I can do it but then, what happens with the go-back arrow?
It is ok when you are in the first folder but when you go to another one, the go-back arrow shows up. Here it's an example of an iPad.

![image-1](https://user-images.githubusercontent.com/8929738/46949308-41513600-d081-11e8-96d6-bd804fb73987.jpg)

Reply:
You're right, of course. Let's go for the `Move here` button solution (+ taking care of contentInsets) then.

Reply:
I didn't noticed these things before, thanks!
I followed your suggestion and here it's the result:

![screen shot 2018-10-17 at 10 15 08](https://user-images.githubusercontent.com/8929738/47072222-e8ef7500-d1f5-11e8-9168-163d7f974407.png)

Reply:
Thanks! Looks great!

Reply:
> whether actions (via row actions, action dots and drag and drop) should be available inside ClientDirectoryPickerViewController. Is it more useful or more confusing?

I think this picker view is only going to be prompted in this two actions:
 1. Select the folder in which you want to move files/folders
 2. Select the folder you want to be the default folder for automatic uploads 

There could be more use cases for this screen in the future but I tend to think it as an isolated view and feature. I'll prefer to keep it as simple as possible and focused in the _pick a folder_ task. 

While developing this feature something else came to my mind. This class is a `ClientQueryViewController` subclass, but eventually, `ClientQueryViewController` will gain more features and this would lead to overriding methods to make the picker do nothing on some actions or use cases.

I think we could make a base class with only a list of files and the sort bar/search bar and then through subclassing, we can make the `ClientQueryViewController` with the multiselection feature, drag and drop support,  3 dots in a cell for actions. And make other class like `DirectoryPickerViewController` with a filter inside for only showing folders and no extra thing but the `Move here` button.

> whether we should add a validation block property to ClientDirectoryPickerViewController, which can be used to "validate" the currently viewed directory, so that f.ex. "Move here" is disabled when the user would otherwise try to move a folder into itself - or to the location it already resides in. Right now, if a user does this, an error will pop up after tapping "Move here".

I just had a quick look at what other apps do (Dropbox) and they allow you to select the current folder as the destination for the move action. Ther prompt the user with an alert saying _Can't move to the same folder_. We can follow this way or we can compare if the `query.rootItem` is the same that the item path we're trying to move and if this is true, disable the _move here_ button

Reply:
@pablocarmu 

> I think we could make a base class with only a list of files and the sort bar/search bar and then through subclassing, we can make the `ClientQueryViewController` with the multiselection feature, drag and drop support, 3 dots in a cell for actions. And make other class like `DirectoryPickerViewController` with a filter inside for only showing folders and no extra thing but the `Move here` button.

I think this is reasonable, but may reach its limits if f.ex. two subclasses should share one additional feature (only that one, not sharing any others). In that case, we'd have to subclass the base class and then subclass that subclass again, ending up with a lot of classes.

An alternative could be to introduce modes to `ClientQueryViewController`, so that code can be shared between modes very selectively. Also, `DirectoryPickerViewController` would no longer be needed and just become a mode of `ClientQueryViewController`. The code would then stay in one place, which might make it easier to navigate long-term, too.

> I just had a quick look at what other apps do (Dropbox) and they allow you to select the current folder as the destination for the move action. Ther prompt the user with an alert saying _Can't move to the same folder_. We can follow this way or we can compare if the `query.rootItem` is the same that the item path we're trying to move and if this is true, disable the _move here_ button

Thanks for checking! It all boils down to "prevent impossible actions from the start" vs. "allow triggering impossible actions and then explain why they aren't possible".

The benefit of the latter is that nobody will end up thinking that "move doesn't work (because the move button is disabled)".